src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageVulnerabilityService.cs

```diff
@@ -17,26 +17,36 @@ namespace NuGet.PackageManagement.VisualStudio
     public class PackageVulnerabilityService : IPackageVulnerabilityService
     {
         private readonly IEnumerable<SourceRepository> _sourceRepositories;
-        private GetVulnerabilityInfoResult _vulnerabilities;
+        private Task<GetVulnerabilityInfoResult> _vulnerabilitiesTask;
         private INuGetUILogger _logger;
+        private readonly object _lock = new();
 
         public PackageVulnerabilityService(IEnumerable<SourceRepository> sourceRepositories, INuGetUILogger logger)
         {
             _sourceRepositories = sourceRepositories ?? throw new ArgumentNullException(nameof(sourceRepositories));
-            _logger = logger ?? throw new ArgumentNullException(nameof(logger)); ;
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
         }
 
         public async Task<List<PackageVulnerabilityMetadataContextInfo>> GetVulnerabilityInfoAsync(PackageIdentity packageId, CancellationToken cancellationToken)
         {
-            _vulnerabilities = await GetAllVulnerabilityDataAsync(cancellationToken);
+            if (_vulnerabilitiesTask == null)
+            {
+                lock (_lock)
+                {
+                    _vulnerabilitiesTask ??= GetAllVulnerabilityDataAsync(cancellationToken);
```

Comment:
`_vulnerabilitiesTask` can be non-null when you enter the lock and you overwrite it, another null check is needed within the lock.